### PR TITLE
Re implement with event processor host

### PR DESCRIPTION
**Pull request by jakelandis**
This is a work in progress PR to help illustrate the proposed changes described on #2

### CODE REVIEW

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+    params['event_hubs'].each_with_index do |event_hub, index|
+      global_config = {}
+      params.each do |k, v|
+        if !k.eql?('id') && !k.eql?('event_hubs')
+          global_config[k] = v
+        end
+      end
+      if event_hub.is_a?(Hash)
+        event_hub.each do |hub, config|
+          config.each do |k, v|
```
Comment (brahmnes) [inline lib/logstash/inputs/azure_event_hubs.rb]: Two things I don't understand about this block:
1. Using your example from line 27:

+  # [
+  #   "event_hub_name1" => {
+  #      "event_hub_connection" => "Endpoint=sb://example1..."
+  #  },
+  #  "event_hub_name2" => {
+  #      "event_hub_connection" => "Endpoint=sb://example2..."
+  #  },
+  # ]

if event_Hub is a hash, then event_hub.each should already be looping on its key/value pairs. Why do you need to do config.each again? 

2. Why are you changing global_config inside this loop? Wouldn't this cause side-effects in adding irrelevant configs for all subsequent eventhubs?

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+      puts "********************************"
+      puts event_hub_name
+      puts config.to_s
+      puts "********************************"
+      event_hub_threads << Thread.new do
+        begin
+          thread = Thread.current
+          java.lang.Thread.currentThread.setName(event_hub_name + "-main")
+          named_thread_factory = LogStash::Inputs::Azure::NamedThreadFactory.new(event_hub_name + "-worker")
+          scheduled_executor_service = Executors.newScheduledThreadPool(8, named_thread_factory)
```
Comment (brahmnes) [inline lib/logstash/inputs/azure_event_hubs.rb]: How do we determine the number 8 here used by the thread pool?
Ideally would it be based on the number of partitions for this eventhub?

File: lib/logstash/inputs/azure_event_hubs.rb (hunk 1 above)
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: This section of code (as well as others) is still under construction... FWIW, I am looping through the config here so that I can convert the sensitive configuration to a Password type which is safe from echo'ing sensitive data to the logs. It's not working yet, the loop is off and i shouldn't merge per hub config up to the global config. 

I should get this (and other parts under construction) sorted out in the few days as well as some testing. 

I put this code up as PR early to help show direction and something a bit more tangible then the descriptions on https://github.com/logstash-plugins/logstash-input-azure_event_hubs/issues/2.  You will likely find many bugs in the code as-is today, but hopefully it illustrates the direction I am going.

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+
+
+    # this blocks the input
+    while !stop?
+      Stud.stoppable_sleep(1) {stop?}
+    end
+
+    # ensure clean shutdown since we own the executor service
+    event_hub_threads.each do |thread|
+      thread[:executor_service].each do |service|
```
Comment (brahmnes) [inline lib/logstash/inputs/azure_event_hubs.rb]: Each thread has only one ScheduledExecutorService and I assume you want to call shutdown on it. What does calling .each on ScheduledExecutorService gives you?

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+  # Used to persists the offsets between restarts and ensure that multiple instances of Logstash process different partitions
+  # This is *stongly* encouraged to be set for production environments.
+  # When this value is set, restarts will pick up from where it left off. Without this value set the initial_position is *always* used.
+  # Alternatively, this configuration may be also be expressed per event hub
+  # [
+  #   "event_hub_name1" => {
+  #      "storage_connection" => "DefaultEndpointsProtocol=https;AccountName=example...."
+  #  }
+  # ]
+  config :storage_connection, :validate => :string, :required => false # TODO: validate as password
```
Comment (brahmnes) [inline lib/logstash/inputs/azure_event_hubs.rb]: You may want to check if EventProcessorHost accept sas key token instead of the storage connection string. From a quick look at the documentation I don't see such option, but you can take another look.

File: lib/logstash/inputs/azure_event_hubs.rb (hunk 2 above)
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: I've since changed this to be an executor per instance of the plugin, configurable, but with a default of 4.  These threads compete with many others in Logstash, so being a bit conservative.   Based solely on observation and experience, reading the data from the EventHub won't be the bottleneck in the processing. However, there should be enough knobs to twist to optimize if needed.

File: lib/logstash/inputs/azure_event_hubs.rb (hunk 3 above)
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: yeah, this was a bug.  I have since moved this logic out to an executor per input instance.

**Comment by jakelandis** [comment]
closing and opening to trigger travis build

**Comment by jakelandis** [comment]
@brahmnes @jsvd - the code should be ready for review. I have a few more edge cases to test, documentation and other tidbits to clean up...  but the bulk of the functionality is there and working. 

I have included some tooling to help with manual integration testing, see the readme in ci/integration. 

Any feedback is appreciated.

**Comment by milo720** [comment]
Firstly, thank you for this, this would bring this plugin into a state much more suited to enterprise solutions. 
Secondly I was wondering at the timescales on this getting through to the live plugin. @brahmnes @jsvd  The pull request looks pretty complete but I'm not sure what the process is for releasing a new version. Thanks again this could be of serious help to me.  @jakelandis

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+  # This plugin supports two styles of configuration
+  # BASIC - You supply a list of Event Hub connection strings complete with the 'EntityPath' that defines the Event Hub name. All other configuration is shared.
+  # ADVANCED - You supply a list of Event Hub names, and under each name provide that Event Hub's configuration. Most all of the configuration options are identical as the BASIC model, except they are configured per Event Hub.
+  # Defaults to BASIC
+  # Example:
+  # azure_event_hubs {
+  #    config_mode => "BASIC"
+  #    event_hub_connections => ["Endpoint=sb://example1...;EntityPath=event_hub_name1"  , "Endpoint=sb://example2...;EntityPath=event_hub_name2"  ]
+  # }
+  config :config_mode, :validate => ['BASIC', 'ADVANCED', 'basic', 'advanced'], :default => 'BASIC'
```
Comment (jakelandis) [inline lib/logstash/inputs/azure_event_hubs.rb]: I struggled alot with how to express this ... and while  I _could_ achieve this separation without this flag it makes the documentation much harder to explain.  The doc still requires a way to describe the different styles and IMO basic vs. advanced makes sense , and it helps to have an actual option rather then just some concept... and you only really need to set it 'advanced' mode.

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+  config :event_hubs, :validate => :array, :required => true # only required for advanced mode
+
+  # BASIC MODE ONLY - The Event Hubs to read from. This is a list of Event Hub connection strings that includes the 'EntityPath'.
+  # All other configuration options will be shared between Event Hubs.
+  # Example:
+  # azure_event_hubs {
+  #    config_mode => "BASIC"
+  #    event_hub_connections => ["Endpoint=sb://example1...;EntityPath=event_hub_name1"  , "Endpoint=sb://example2...;EntityPath=event_hub_name2"  ]
+  # }
+  config :event_hub_connections, :validate => :array, :required => true # only required for basic mode
```
Comment (jakelandis) [inline lib/logstash/inputs/azure_event_hubs.rb]: i started down the path (as-is in this review)

* event_hubs = advanced mode = key of hash
* event_hub_connections = basic mode = array of connections

... however reconsidering that this ... it was an attempt to make things simpler for the basic v.s. advanced divide, but it may actually make things worse. 

I am considering changing this to : 

* event_hubs in advanced mode = key of hash
* event_hubs in basic mode = array of connections
* remove event_hub_connections configuration option

Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: arguing with myself here ....

I am back to advocating the way it is in this review.... When working with this plugin via a module, where the configuration is expressed via YAML, this does indeed make it more clear, especially when this option is set from the command line. 

I am still open to discussion for this change but we will also need to account for the user's experience configuring this via YAML or Logstash config.  I can elaborate on the YAML config if needed.

Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: @milo720 - thanks for your feed back. 

> could elaborate on the yaml

I am building a Logstash [module](https://www.elastic.co/blog/monitoring-azure-with-the-elastic-stack-feedback-welcome) that consumes this plugin.  I am trying to keep the module requirements separate from the plugin, since the plugin is a standalone entity.  However, since it is green field(ish) for both, I am trying to ensure the best user experience for both. 

This means that I need to take into account 2 ways of configuring this plugin.
1) Via the normal Logstash configuration
2) Via the module YAML conventions (see [here](https://www.elastic.co/guide/en/logstash/current/logstash-modules.html) for Module based config)

To complicate matters, this plugin also has
a) 'Basic' 
b) 'Advanced' 
modes of configuration. 

This configuration is split into basic (default) and advanced to help keep easy things easy, but also allow for more powerful configuration, but also causes a 1a, 1b, 2a, 2b ways of expressing the configuration. There is also a 5th way to configure via modules and the command line, which is like a 3a means of configuration.  

A full example (which will be in the docs) can help explain.

Basic configuration
-----------
Basic configuration (1a) can look like this:
```ruby
input {
   azure_event_hubs {
      threads => 8
      decorate_events => true
      consumer_group => "logstash"
      storage_connection => "DefaultEndpointsProtocol=https;AccountName=example...."
      event_hub_connections => ["Endpoint=sb://example1...EntityPath=insights-logs-errors", "Endpoint=sb://example2...EntityPath=insights-metrics-pt1m"]
   }
}
```
Same basic configuration (2a) via the Module can look like this:
```yaml
modules:
  - name: azure
    var.input.azure_event_hubs.threads: 8
    var.input.azure_event_hubs.decorate_events: true
    var.input.azure_event_hubs.consumer_group: "logstash"
    var.input.azure_event_hubs.storage_connection: "DefaultEndpointsProtocol=https;AccountName=example...."
    var.input.azure_event_hubs.event_hub_connections:
      - "Endpoint=sb://example1...EntityPath=insights-logs-errors"
      - "Endpoint=sb://example2...EntityPath=insights-metrics-pt1m"
```

Same basic configuration (3a) via Module command line can look like this:
```bash
bin/logstash --modules azure -M "azure.var.elasticsearch.host=es.mycloud.com" -M "azure.var.input.azure_event_hubs.threads=8" -M "azure.var.input.azure_event_hubs.consumer_group=logstash" -M "azure.var.input.azure_event_hubs.decorate_events=true" -M "azure.var.input.azure_event_hubs.event_hub_connections=Endpoint=sb://example1...EntityPath=insights-logs-errors" -M "azure.var.input.azure_event_hubs.storage_connection=DefaultEndpointsProtocol=https;AccountName=example...."
```
All the above example express the same thing. Monitor 2 event hubs with different connection strings, everything else is the same. 

Advanced configuration 
-----------
Advanced configuration (1b) can look like this:
```ruby
input {
   azure_event_hubs {
      threads => 8
     consumer_group => "logstash"
     config_mode => "ADVANCED"
     event_hubs => [
        {"insights-operational-logs" => {
         event_hub_connection => "Endpoint=sb://example1..."
         storage_connection => "DefaultEndpointsProtocol=https;AccountName=example1...."
         initial_position => "HEAD"
        decorate_events => "true"
        }},
      {"insights-metrics-pt1m" => {
         event_hub_connection => "Endpoint=sb://example2..."
         storage_connection => "DefaultEndpointsProtocol=https;AccountName=example2...."
         initial_position => "TAIL"
        decorate_events => "true"
       }},
      {"insights-logs-errors" => {
        event_hub_connection => "Endpoint=sb://example3..."
        storage_connection => "DefaultEndpointsProtocol=https;AccountName=example3...."
        initial_position => "LOOK_BACK"
       decorate_events => "false"
      }},
      {"insights-operational-logs" => {
        event_hub_connection => "Endpoint=sb://example4..."
        storage_connection => "DefaultEndpointsProtocol=https;AccountName=example4...."
        initial_position => "HEAD"
       decorate_events => "true"
    }}
  ]
 }
}
```
Same advanced configuration (1b) as a Module:
```yaml
modules:
  - name: azure
    var.input.azure_event_hubs.threads: 8
    var.input.azure_event_hubs.decorate_events: true
    var.input.azure_event_hubs.consumer_group: logstash
    var.input.azure_event_hubs.event_hubs:
      - ["name",                      "event_hub_connection",      "storage_connection",                                      "initial_position", "decorate_events"]
      - ["insights-operational-logs", "Endpoint=sb://example1...", "DefaultEndpointsProtocol=https;AccountName=example1....", "HEAD",             "true"]
      - ["insights-metrics-pt1m",     "Endpoint=sb://example2...", "DefaultEndpointsProtocol=https;AccountName=example2....", "TAIL",             "true"]
      - ["insights-logs-errors",      "Endpoint=sb://example3...", "DefaultEndpointsProtocol=https;AccountName=example3....", "LOOK_BACK",        "false"]
      - ["insights-operational-logs", "Endpoint=sb://example4...", "DefaultEndpointsProtocol=https;AccountName=example4....", "HEAD",             "true"]
```
_note advanced config via Module and command line wont be supported

All the above examples allow to read from multiple event hubs, each with their own specific configuration + the 'global' config (in this example the consumer_group). 

_note in all advanced configuration, I need to be careful to allow duplicated names since some of the things that I may need to monitor have in-flexible naming. For example, if wanted to monitor 2 activity logs in 2 different subscriptions, I need to monitor two event hubs named `insights-operational-logs`.  This make the configuration a bit weird since I can't simply use the names as the key. 

Issue in question
---------
If you notice that for advanced you configure `event_hubs`, but for basic you configure `event_hub_connections`.  I had considered pushing it all into just `event_hubs`...but I know think someone will either use basic OR advanced, so as long as there are good example of both, the naming doesn't matter too much.... and it is a bit a weird to have a single configuration option be so wildly type variant. 

Nothing is set in stone yet, any feedback with the complete examples is welcome.

Comment (milo720) [reply lib/logstash/inputs/azure_event_hubs.rb]: Ahh sorry, as you can probably tell I have very little experience with modules and should have probably have read you previous comments in a bit more detail. In terms of the logstash config the full examples makes them very clear (Only just now notices you have a lot of examples below your comment in the code). I think having the two modes is worth it and can't really think of an option that is better. I can see both ways of writing the config being useful and could see many instances of config bloat from repeated config if only advanced was available. In terms of the yaml I have not really use modules so can't be much help there. One thing i was going to ask was can you have several differently named config blocks for one plugin, for example azure_event_hubs_basic and azure_event_hubs_Advanced input blocks, it may solidify the boundary between the different types and would allow the removal of the flag while keeping what type it is using explicit. I don't know how feasible that is, so if it is difficult/ against consistency in the Logstash plugin ecosystem please ignore this suggestion.

One thing that wasn't immediately obvious in the config was the 'Head', 'Tail', 'LOOK_BACK' and I had to look for the clarification in your comments. This is due to the fact that I felt that head and tail could be ambiguous between which end they meant. The is due to the fact different technologies do use them in different ways e.g Git - head is the most recent and Haskell list  - head is the first one on. While you can obviously read the docs maybe a more obvious one such as most recent/ oldest, or something along those lines, would be easier to use.

On a separate note can I ask what you estimate on time scales on this becoming available or use are as I have a project that I want to use this and Logstash in but will not be able to do so if this is not ready soon.
Thanks and feel free to ask my opinion on anything else.
P.S Bit late and quite tiered so please excuse any obvious mistakes.

Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: @milo720 - thanks for the feedback. 

several differently named config blocks for one plugin ... can't do this without alot of ugly hackery.  

head vs. tail ... I see how that can be ambiguous, I will think on this a bit to see if I can make the naming clearer. 

release estimate ... It needs to go through review and doc, and I still have a couple minor things to clean up. I would guess about 2-3 weeks, abit longer if issues are discovered or large changes requested.

**Comment by jakelandis** [comment]
Doc is started at https://github.com/logstash-plugins/logstash-input-azure_event_hubs/blob/doc/docs/index.asciidoc

...and the TODO's will be populated with examples from:
https://github.com/jakelandis/logstash-input-azure_event_hubs/blob/re-implement-with-event-processor-host/lib/logstash/inputs/azure_event_hubs.rb

File: lib/logstash/inputs/azure_event_hubs.rb (hunk 6 above)
Comment (milo720) [reply lib/logstash/inputs/azure_event_hubs.rb]: OK that might be feasible for me. If there is anything I can do to help please let me know. It maybe worth knowing that I have a knowledge of ruby but that is rusty and was never that good in the first place. If you think it could help I would be happy to read over and review what I can, Also I have a somewhat simple Logstash/event hub set at home so if you want to chuck some testing at me feel free.

**Comment by milo720** [comment]
Hi
Is this still on track to be done in the next 2 weeks

**Comment by jakelandis** [comment]
> IMO we could remove the config_mode 

`config_mode` only needs to be set if setting it to advanced. This requires the user to opt-in explicitly to using advanced mode, basic mode is the default and can be left out. IMO keeping the `config_mode` makes it easier to reason about since it is an explicit configuration and the documentation has something tangible to call the different configuration modes.  See https://github.com/logstash-plugins/logstash-input-azure_event_hubs/blob/doc/docs/index.asciidoc  for full examples and explanations. 

I understand the argument about adding technically unnecessary config (which this is...) , however I had a really hard time documenting this without the `config_mode`. 

I am open to changes (even a complete re-think of the options) , but in this particular case, removing `config_mode` could harm documentation and/or reasoning about the options.

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+  # Note - this options is only used when initial_position => "LOOK_BACK"
+  # Value is expressed in seconds, default is 1 day
+  # BASIC Example:
+  # azure_event_hubs {
+  #    config_mode => "BASIC"
+  #    event_hub_connections => ["Endpoint=sb://example1...;EntityPath=event_hub_name1"]
+  #    initial_position => "LOOK_BACK"
+  #    initial_position_look_back => 86400
+  # }
+  # ADVANCED example:
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: Documentation is started here: https://github.com/logstash-plugins/logstash-input-azure_event_hubs/blob/doc/docs/index.asciidoc 
cc @karenzone 

>  I'd suggest documenting only the basic example
I like this suggestion.  It keep the individual options for getting too complicated.  However, I  _may_ keep the documentation in the comments here as-is (they helped me when developing).

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+  #   config_mode => "ADVANCED"
+  #   event_hubs => [
+  #       { "event_hub_name1" => {
+  #           event_hub_connection => "Endpoint=sb://example1..."
+  #           initial_position => "LOOK_BACK"
+  #           initial_position_look_back => 86400
+  #       }}
+  #    ]
+  # }
+  config :initial_position_look_back, :validate => :number, :default => 86400
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: it's ugly .. but how about `initial_position_look_back_period` ?  This only applies to the initial position, e.g. if they use the recommended blob store, this option is only used once.

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+
+      params['event_hubs'].each do |event_hub|
+        raise "event_hubs must be a Hash" unless event_hub.is_a?(Hash)
+        event_hub.each do |event_hub_name, config|
+          config.each do |k, v|
+            if 'event_hub_connection'.eql?(k) || 'storage_connection'.eql?(k) # protect from leaking logs
+              config[k] = ::LogStash::Util::Password.new(v)
+            end
+          end
+          if config['event_hub_connection'] #add the 's' to pass validation
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: It's mandatory.  I played with not making it mandatory, but it got extra confusing.  will update.

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 10]
```diff
@@ -0,0 +1,477 @@
...
+          end
+
+          config.merge!({'event_hubs' => [event_hub_name]})
+          config.merge!(global_config) {|k, v1, v2| v1}
+          @event_hubs_exploded << config
+        end
+      end
+    else # basic config
+      params['event_hubs'] = ['dummy'] # trick the :required validation
+      if params['event_hub_connections']
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: I'll add a test for this, but pretty sure it will just fail validation.

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+      end
+    end
+    @logger.debug("Exploded Event Hub configuration: #{@event_hubs_exploded.to_s}")
+  end
+
+  def run(queue)
+    event_hub_threads = []
+    named_thread_factory = LogStash::Inputs::Azure::NamedThreadFactory.new("azure_event_hubs-worker")
+    scheduled_executor_service = Executors.newScheduledThreadPool(@threads, named_thread_factory)
+    @event_hubs_exploded.each do |event_hub|
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: I'll look into this...

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+            event_processor_host = EventProcessorHost.new(
+                EventProcessorHost.createHostName('logstash'),
+                event_hub_name,
+                event_hub['consumer_group'],
+                event_hub['event_hub_connections'].first.value, #there will only be one in this array by the time it gets here
+                event_hub['storage_connection'].value,
+                event_hub['storage_container'] ? event_hub['storage_container'] : event_hub_name,
+                scheduled_executor_service)
+          else
+            @logger.warn("You have NOT specified a `storage_connection_string` for #{event_hub_name}. This configuration is only supported for a single Logstash instance.")
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: > for the second part of this warning, could we instead call out the impact of starting multiple instances w/o a storate account (duplicate data)?

👍 

> Also, a user will also have problems by starting two instances of the plugin in the same Logstash instance, right?

I don't think so, it should work fine.  Functionally if they define the same event hub/consumer group and/or storage container , it won't read twice, but the same applies to two different LS instances.  I can test, but do you see a problematic area ?

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+            @logger.info(msg) unless event_hub['storage_connection']
+            options.setInitialPositionProvider(EventProcessorOptions::EndOfStreamInitialPositionProvider.new(options))
+          when 'LOOK_BACK'
+            msg = "Configuring Event Hub #{event_hub_name} to read events starting at 'now - #{@initial_position_look_back}' seconds."
+            @logger.debug("If this is the initial read... " + msg) if event_hub['storage_connection']
+            @logger.info(msg) unless event_hub['storage_connection']
+            options.setInitialPositionProvider(LogStash::Inputs::Azure::LookBackPositionProvider.new(@initial_position_look_back))
+          end
+          event_processor_host.registerEventProcessorFactory(LogStash::Inputs::Azure::ProcessorFactory.new(queue, event_hub['codec'].clone, event_hub['checkpoint_interval'], self.method(:decorate), event_hub['decorate_events']), options)
+              .whenComplete {|x, e|
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: I can try, but IIRC the mapping from java lambda requires `{` syntax

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+    # This blocks the input till all the threads have run to completion.
+    event_hub_threads.each do |thread|
+      thread.join
+    end
+
+    # Ensure proper shutdown of executor service. # Note - this causes a harmless warning in the logs that scheduled tasks are being rejected.
+    scheduled_executor_service.shutdown
+    begin
+      scheduled_executor_service.awaitTermination(10, TimeUnit::MINUTES);
+    rescue => e
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: It's catching an interrupt exception. This is the typical pattern to close down an executor service in Java. I thought (ruby) StandardError ==> (java) Exception , is that wrong ?

File: lib/logstash/inputs/azure_event_hubs.rb (hunk 5 above)
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: oops .. commented in main thread instead of inline. 
reply here: https://github.com/logstash-plugins/logstash-input-azure_event_hubs/pull/3#issuecomment-397316069

File: lib/logstash/inputs/azure_event_hubs.rb [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+        @logger.info("Event Hub #{event_hub_name} is initializing... ")
+        begin
+          if event_hub['storage_connection']
+            event_processor_host = EventProcessorHost.new(
+                EventProcessorHost.createHostName('logstash'),
+                event_hub_name,
+                event_hub['consumer_group'],
+                event_hub['event_hub_connections'].first.value, #there will only be one in this array by the time it gets here
+                event_hub['storage_connection'].value,
+                event_hub['storage_container'] ? event_hub['storage_container'] : event_hub_name,
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply lib/logstash/inputs/azure_event_hubs.rb]: fetch

File: lib/logstash/inputs/azure_event_hubs.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (jsvd) [reply lib/logstash/inputs/azure_event_hubs.rb]: yeah, I wasn't sure about the exception hierarchy there. As long as the exception inherits from java.lang.Exception it's ok (but not for java.lang.Error, you need to catch Exception for that)

File: lib/logstash/inputs/azure_event_hubs.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (jsvd) [reply lib/logstash/inputs/azure_event_hubs.rb]: No, it was a misunderstanding on my part 👍

File: logstash-input-azure_event_hubs.gemspec [hunk 16]
```diff
@@ -0,0 +1,28 @@
+GEM_VERSION = File.read(File.expand_path(File.join(File.dirname(__FILE__), "VERSION"))).strip unless defined?(GEM_VERSION)
+
+Gem::Specification.new do |s|
+  s.name          = 'logstash-input-azure_event_hubs'
+  s.version       = GEM_VERSION
+  s.licenses      = ['Apache-2.0']
+  s.summary       = 'Consumes events from Azure Event Hubs for use with Logstash'
+  s.description   = 'This gem is a Logstash plugin required to be installed on top of the Logstash core pipeline using $LS_HOME/bin/logstash-plugin install gemname. This gem is not a stand-alone program'
+  s.homepage      = 'http://www.elastic.co/guide/en/logstash/current/index.html"'
```
Comment (milo720) [inline logstash-input-azure_event_hubs.gemspec]: The extra quotes caused an error for my while building the gem

File: README.md [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+#### Code
+- To get started, you'll need JRuby with the Bundler gem installed.
+
+- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).
+
+- Install dependencies
+```sh
+./gradlew vendor
+bundle install
+```
```
Comment (milo720) [inline README.md]: Hi guys maybe i', doing this wrong but i download the repo and run `gradlew.bat vendor` (windows) and I get.
A problem occurred evaluating root project 'logstash-input-azure_event_hubs-re-implement-with-event-processor-host'.
> Could not find method baseFeature() for arguments [true] on root project 'logstash-input-azure_event_hubs-re-implement-with-event-processor-host' of type org.gradle.api.Project.

Comment (milo720) [reply README.md]: Gradlew worked instantly on my Linux partition, so it could either be windows or something in my environment (very possible due to other dev work).

**Comment by milo720** [comment]
Have everything up and running on linux after quite a while install and messing with the path. Will start running through the basic manual test cases.

**Comment by milo720** [comment]
Have done the two manual test described, here are my results.
 No storage:
      Added events using producer
      Started LS instances using the docker testing script with the no_storage config
      Look at how many events present using consumer - 1023 (some left over from before)
      Looked at LS1 results and saw 1023 in and out 
      Looked at LS2 results and saw 1023 in and out 
      Stopped and Started Logstash again 
      Looked at LS1 results and saw 1023 in and out 
      Looked at LS2 results and saw 1023 in and out 
with storage:
      Run consumer find 1022 events in hub.
      Started LS instances using test docker script and the with_storgae config
      LS1 shows 1022 events in and out
      LS2 shows 0 events in and out
      Stop and then start the LS instances
      LS1 shows 0 events in and out
      LS2 shows 0 events in and out
Notes:
  - I had to run the script using sudo for docker to work so I added the setting of the environment variables into the script.
  - I found it odd that I lost and event between the two tests but perhaps one of the older event expired between the two test.
  - I have also done a small amount of exploratory testing (playing around) and everything worked as expected.

These test look to have passed, I will probably get some time tomorrow night (BST) night so if you have any tests you want me to run please give me a shout @jsvd @jakelandis

File: README.md (hunk 17 above)
Comment (jakelandis) [reply README.md]: thanks for trying this out , i will give it a go on my Windows box to make sure it works there too.

**Comment by jakelandis** [comment]
@milo720 - thanks for helping to test. That 1 event difference is odd, I'll try to reproduce. Also I noticed that when reading an event hub for the first time with a storage account, the events in the event hub are not evenly balanced between logstash instances (in your case all were picked up by 1 instance).  However, once data starts stream in, the load starts to get balanced between instances.

**Comment by milo720** [comment]
I will give it another try tonight as well, and clear my event hub before.

**Comment by jakelandis** [comment]
Review changes are in and ready for another round of reviews : https://github.com/logstash-plugins/logstash-input-azure_event_hubs/pull/3/commits/55a3fc6168b4ee78b57d88b57fb4c89e29ad6792 and https://github.com/logstash-plugins/logstash-input-azure_event_hubs/pull/3/commits/8965bf93bb8e24899f880ba6b0bf9f7ac8c7a809

File: README.md (hunk 17 above)
Comment (jakelandis) [reply README.md]: I fixed some forward vs. backslash issue in the gradle task. It should build on Windows now.

**Comment by milo720** [comment]
Tested building and then running the plugin locally in windows and all seems good, events were read down as expected.

Also checked that if a logstash with storage is taken offline it continues where it left off. Did this using a script with logs with increasing numbers and checked that all the numbers from 1-1000 where read only once. Have yet to try the maven or docker bits on windows.

**Comment by jsvd** [comment]
I've ran a few tests and I was able to read data correctly from different event hubs.
I think we're good to merge this, the only thing missing is linking `rake vendor` to `gradle vendor`, since most of our deployment tooling relies on `bundle install && bundle exec rake vendor && bundle exec rspec && bundle exec rake publish`

**Comment by jakelandis** [comment]
`rake vendor` added as alias to `gradle vendor` and all commits squashed down to one.

**Comment by jakelandis** [comment]
@jsvd / @milo720 - thanks for the help and reviews.  merging ...
